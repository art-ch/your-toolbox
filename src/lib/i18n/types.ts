export type DefaultLanguage = 'en';

export type Language = DefaultLanguage | 'uk' | 'ru';

export type TranslationNamespace =
  | 'common'
  | 'homePage'
  | 'footer'
  | 'time'
  | 'asceticExerciseDuration';
