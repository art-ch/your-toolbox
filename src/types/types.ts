export type LocalHref = `/${string}`;
export type GlobalHref = `https://${string}`;
export type Href = LocalHref | GlobalHref;

export type AuthorMentionType = {
  name: string;
  href: GlobalHref;
};
