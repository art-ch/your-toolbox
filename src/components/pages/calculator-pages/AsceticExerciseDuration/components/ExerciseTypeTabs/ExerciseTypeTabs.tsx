import React from 'react';

import { Tabs, TabsList, TabsContent } from '@/components/ui/tabs';
import { FormTab } from './components/FormTab';
import { Trigger } from './components/Trigger';
import { EXERCISE_TYPE_TABS_CONFIG } from './config/tabsConfig';
import { ExerciseTypeConfigItem } from './ExerciseTypeTabs.types';
import { useExerciseTypeUrlState } from './hooks/useExerciseTypeUrlState';

export const ExerciseTypeTabs = () => {
  const { exerciseTypePath, handleTabChange } = useExerciseTypeUrlState(
    EXERCISE_TYPE_TABS_CONFIG
  );

  const renderTabs = (
    config: ExerciseTypeConfigItem[],
    depth: number = 0
  ): React.ReactNode => {
    const currentValue = exerciseTypePath[depth];
    const defaultValue = currentValue || config[0]?.key || '';

    return (
      <Tabs
        value={defaultValue}
        onValueChange={(value) => handleTabChange(value, depth)}
      >
        <TabsList className="my-0 mx-auto bg-white">
          {config.map((configItem) => (
            <Trigger
              key={configItem.key}
              value={configItem.key}
              color={configItem.themeColor}
              icon={configItem.icon}
            />
          ))}
        </TabsList>

        {config.map((configItem) => {
          // If item has forms, render them directly
          if (configItem.forms && configItem.forms.length > 0) {
            return (
              <FormTab key={configItem.key} value={configItem.key}>
                {configItem.forms.map((FormComponent, index) => (
                  <FormComponent key={index} />
                ))}
              </FormTab>
            );
          }

          // If item has children, render nested tabs
          if (configItem.children && configItem.children.length > 0) {
            return (
              <TabsContent key={configItem.key} value={configItem.key}>
                {renderTabs(configItem.children, depth + 1)}
              </TabsContent>
            );
          }

          // Fallback for items without forms or children
          return null;
        })}
      </Tabs>
    );
  };

  return renderTabs(EXERCISE_TYPE_TABS_CONFIG);
};
