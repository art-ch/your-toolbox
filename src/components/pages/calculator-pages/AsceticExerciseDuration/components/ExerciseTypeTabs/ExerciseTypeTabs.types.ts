export enum ThemeColor {
  RED = 'red',
  INDIGO = 'indigo',
  GREEN = 'green',
  CYAN = 'cyan'
}

export type ExerciseTypeConfigItem = {
  key: string;
  themeColor: ThemeColor;
  icon: React.ReactNode;
  forms?: React.ComponentType[];
  children?: ExerciseTypeConfigItem[];
};

export type ExerciseTypeConfig = ExerciseTypeConfigItem[];
