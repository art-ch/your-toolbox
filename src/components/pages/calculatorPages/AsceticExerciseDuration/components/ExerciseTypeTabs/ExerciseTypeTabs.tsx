import React from 'react';

import { Tabs, TabsTrigger, TabsList, TabsContent } from '@/components/ui/tabs';
import { FaPerson, FaPersonRunning, FaPersonSwimming } from 'react-icons/fa6';
import { GiMeditation } from 'react-icons/gi';

import { AsceticExerciseDurationSection } from '../AsceticExerciseDurationSection';
import {
  CalculateCleansingCyclesForm,
  CalculateSpeedForm,
  CalculateTotalTimeForm
} from '../DynamicExerciseForms';
import {
  CalculateDurationFromLayersForm,
  CalculateExposureTimeForm,
  CalculateLayersFromDurationForm,
  CalculateMentalLayersCleansedForm,
  CalculateTemperatureForm
} from '../StaticExerciseForms';

export const ExerciseTypeTabs = () => (
  <Tabs defaultValue="dynamic">
    <TabsList className="my-0 mx-auto">
      <TabsTrigger value="dynamic">
        <FaPersonRunning />
        Dynamic
      </TabsTrigger>
      <TabsTrigger value="static">
        <FaPerson />
        Static
      </TabsTrigger>
    </TabsList>
    <TabsContent value="dynamic">
      <AsceticExerciseDurationSection>
        <CalculateCleansingCyclesForm />
        <CalculateSpeedForm />
        <CalculateTotalTimeForm />
      </AsceticExerciseDurationSection>
    </TabsContent>
    <TabsContent value="static">
      <Tabs defaultValue="static_ground">
        <TabsList className="my-0 mx-auto">
          <TabsTrigger value="static_ground">
            <GiMeditation />
            Ground
          </TabsTrigger>
          <TabsTrigger value="static_water">
            <FaPersonSwimming />
            Water
          </TabsTrigger>
        </TabsList>
        <TabsContent value="static_ground">
          <AsceticExerciseDurationSection>
            <CalculateDurationFromLayersForm />
            <CalculateLayersFromDurationForm />
          </AsceticExerciseDurationSection>
        </TabsContent>
        <TabsContent value="static_water">
          <AsceticExerciseDurationSection>
            <CalculateMentalLayersCleansedForm />
            <CalculateTemperatureForm />
            <CalculateExposureTimeForm />
          </AsceticExerciseDurationSection>
        </TabsContent>
      </Tabs>
    </TabsContent>
  </Tabs>
);
