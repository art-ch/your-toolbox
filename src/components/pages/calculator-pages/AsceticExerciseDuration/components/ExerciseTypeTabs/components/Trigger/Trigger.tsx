import { TabsTrigger } from '@/components/ui/tabs';
import { useTranslation } from 'react-i18next';
import { ThemeColor } from '../../ExerciseTypeTabs.types';
import { THEME_COLOR_TO_CLASS_MAP } from '../../config/themeColorToClassMap';

export type TriggerProps = {
  icon: React.ReactNode;
  color: ThemeColor;
  value: string;
};

export const Trigger = ({ value, color, icon }: TriggerProps) => {
  const { t } = useTranslation('asceticExerciseDuration');
  const colorClasses = THEME_COLOR_TO_CLASS_MAP[color];

  return (
    <TabsTrigger
      value={value}
      className={`cursor-pointer mx-1 p-4 gap-x-2 ${colorClasses.text} ${colorClasses.activeBg} data-[state=active]:text-white`}
    >
      {icon}
      {t(value)}
    </TabsTrigger>
  );
};
