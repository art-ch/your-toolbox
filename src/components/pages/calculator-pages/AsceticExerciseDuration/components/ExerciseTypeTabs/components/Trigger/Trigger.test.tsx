import { render, screen } from '@testing-library/react';
import { Tabs, TabsList } from '@/components/ui/tabs';
import { Trigger } from './Trigger';
import { ThemeColor } from '../../ExerciseTypeTabs.types';

// Mock react-i18next
jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key
  })
}));

describe('Trigger', () => {
  const mockIcon = <span data-testid="test-icon">üèÉ</span>;
  const defaultProps = {
    icon: mockIcon,
    color: ThemeColor.RED,
    value: 'test-exercise'
  };

  it('should render', () => {
    const { container } = render(
      <Tabs value="test-exercise">
        <TabsList>
          <Trigger {...defaultProps} />
        </TabsList>
      </Tabs>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with correct props', () => {
    render(
      <Tabs value="test-exercise">
        <TabsList>
          <Trigger {...defaultProps} />
        </TabsList>
      </Tabs>
    );

    expect(screen.getByText('test-exercise')).toBeInTheDocument();
    expect(screen.getByTestId('test-icon')).toBeInTheDocument();
  });

  it('should pass correct value to TabsTrigger', () => {
    const { container } = render(
      <Tabs value="custom-value">
        <TabsList>
          <Trigger {...defaultProps} value="custom-value" />
        </TabsList>
      </Tabs>
    );

    const trigger = container.querySelector('[data-state="active"]');
    expect(trigger).toBeInTheDocument();
    // The value is passed as a prop to TabsTrigger, not as an attribute
    expect(trigger).toHaveTextContent('custom-value');
  });

  it('should handle different icon types', () => {
    const customIcon = <div data-testid="custom-icon">Custom Icon</div>;

    render(
      <Tabs value="test-exercise">
        <TabsList>
          <Trigger {...defaultProps} icon={customIcon} />
        </TabsList>
      </Tabs>
    );

    expect(screen.getByTestId('custom-icon')).toBeInTheDocument();
    expect(screen.queryByTestId('test-icon')).not.toBeInTheDocument();
  });

  it('should render with different values', () => {
    const { rerender } = render(
      <Tabs value="exercise1">
        <TabsList>
          <Trigger {...defaultProps} value="exercise1" />
        </TabsList>
      </Tabs>
    );

    expect(screen.getByText('exercise1')).toBeInTheDocument();

    rerender(
      <Tabs value="exercise2">
        <TabsList>
          <Trigger {...defaultProps} value="exercise2" />
        </TabsList>
      </Tabs>
    );

    expect(screen.getByText('exercise2')).toBeInTheDocument();
  });
});
