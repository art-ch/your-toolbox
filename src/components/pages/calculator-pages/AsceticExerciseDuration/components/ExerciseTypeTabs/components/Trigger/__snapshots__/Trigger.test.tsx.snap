// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trigger should render 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
    data-orientation="horizontal"
    data-slot="tabs"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="bg-muted text-muted-foreground inline-flex h-9 w-fit items-center justify-center rounded-lg p-[3px]"
      data-orientation="horizontal"
      data-slot="tabs-list"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-«r0»-content-test-exercise"
        aria-selected="true"
        class="dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 cursor-pointer mx-1 p-4 gap-x-2 text-red-500 data-[state=active]:bg-red-500 data-[state=active]:text-white"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="active"
        id="radix-«r0»-trigger-test-exercise"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          data-testid="test-icon"
        >
          🏃
        </span>
        test-exercise
      </button>
    </div>
  </div>
</div>
`;
